filters/ratelimit/ratelimit.go

```diff
@@ -15,24 +16,51 @@ import (
 	"github.com/zalando/skipper/ratelimit"
 )
 
-// RetryAfterKey is used as key in the context state bag
-const RetryAfterKey = "#ratelimitretryafter"
-
-// RouteSettingsKey is used as key in the context state bag
-const RouteSettingsKey = "#ratelimitsettings"
-
 type spec struct {
 	typ        ratelimit.RatelimitType
+	provider   RatelimitProvider
 	filterName string
 }
 
 type filter struct {
 	settings ratelimit.Settings
+	provider RatelimitProvider
+}
+
+// RatelimitProvider returns a limit instance for provided Settings
+type RatelimitProvider interface {
```

Comment:
I am not completely convinced about the public interface with a private method, but I see the problem and the reason.

Alternatively:

1. we could move the RatelimitProvider interface to the ratelimit package (from filters/ratelimit)
2. the interface could have a public GetRatelimit method, returning an interface with the same methods, or better the necessarily required ones, as the ratelimit.Ratelimit structure, called e.g. Ratelimiter
3. the ratelimit.Registry would implement the ratelimit.RatelimitProvider interface, and would mark the current Get method deprecated

Then we could still conveniently stub the provider. WDYT?

Reply:
At first I have tried to adjust existing interfaces from `ratelimit` package but quickly fell into a tarpit. As I am and was not sure about deprecation and compatibility policy I have decided to workaround it in `filters/ratelimit` package.

The only downside of _sealed_ `RatelimitProvider` interface I see is that it prevents users from creating ratelimit filters from _outside_ of the `filters/ratelimit` package _and without_ `ratelimit.Registry` at hand. If that is needed (not sure why) it could be achieved by having an extra interface and making `get` public (though registry adapter would still be needed).

There is another option without new interfaces at all - make filters require something like `func CheckLimit(*http.Request, *Settings) (retryAfter int)` and create registry adapter for it.

The interface mismatch between what filters need and what ratelimit package provides feels to big.

I would like to address it later if and when we see this change has chances to land into mainline (i.e. there are no showstoppers).

Reply:
well, ok, though it still scratches my aesthetics a bit :smile: